src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -707,7 +707,7 @@ void ShenandoahEvacUpdateCleanupOopStorageRootsClosure::do_oop(oop* p) {
       }
       ShenandoahHeap::atomic_update_oop(resolved, p, obj);
       assert(_heap->cancelled_gc() ||
-             _mark_context->is_marked(resolved) && !_heap->in_collection_set(resolved),
+             (_mark_context->is_marked(resolved) && !_heap->in_collection_set(resolved)),
```

Comment:
The change for this PR seems fine, but the whole assert looks weird, and lacks richer diagnostics Shenandoah asserts do. I believe the forwardee is explicitly marked, given it is a new copy after TAMS. So the whole thing might be just:

```
  shenandoah_assert_not_in_cset_except(p, resolved, _heap->cancelled_gc());
```

Right, @zhengyu123? We should do that in separate PR.

Reply:
Thanks.  Removed as part of merge conflict resolution.